File: Remove Linked List Elements/Remove_Linked_List_Elements.py
# Runtime: 68 ms, faster than 100.00% of Python3 online submissions for Remove Linked List Elements.
# Memory Usage: 15.9 MB, less than 28.32% of Python3 online submissions for Remove Linked List Elements.

# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution:
    def removeElements(self, head: 'ListNode', val: 'int') -> 'ListNode':
        
        dummyHead = ListNode(0)
        dummyHead.next = head

        preNode = dummyHead
        while head:
            if head.val == val:
                preNode.next = head.next
                del head
                head = preNode.next
            else:
                preNode = head
                head = head.next
        
        return dummyHead.next

File: Remove Linked List Elements/Remove_Linked_List_Elements.cpp
// 最简单的链表的题目 纯粹是用来练手的
// 对于这种链表的题目来说，新建一个哑节点有时候是非常方便的

// Runtime: 36 ms, faster than 81.62% of C++ online submissions for Remove Linked List Elements.
// Memory Usage: 11.6 MB, less than 11.82% of C++ online submissions for Remove Linked List Elements.

/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */

class Solution 
{
public:
    ListNode* removeElements(ListNode* head, int val) 
    {
        // 建立一个哑节点
        ListNode dummyHead(0);
        dummyHead.next = head;

        for (ListNode *preNode = &dummyHead; head;)
        {
            if (head->val == val)
            {
                preNode->next = head->next;
                delete head;
                head = preNode->next;
            }
            else
            {
                preNode = head;
                head = head->next;
            }
        }
        return dummyHead.next;
    }
};